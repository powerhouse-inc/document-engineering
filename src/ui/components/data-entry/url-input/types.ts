import type { DiffMode, WithDifference } from "../../../../scalars/components/types.js";
import { IconName } from "../../icon/index.js";
import type { InputBaseProps } from "../../../../scalars/components/types.js";

interface UrlInputWithDifference
  extends Omit<WithDifference<string>, "diffMode"> {
  diffMode?: Extract<DiffMode, "sentences">;
}

type PlatformIcon = IconName | React.ReactElement;

interface UrlInputProps
  extends UrlInputWithDifference, InputBaseProps<string>,
    Omit<
      React.InputHTMLAttributes<HTMLInputElement>,
      "pattern" | "value" | "defaultValue" | "name" | "maxLength"
    > {
  showWarnings?: boolean;
  platformIcons?: Record<string, PlatformIcon>;
}

export type { UrlInputProps, PlatformIcon, UrlInputWithDifference };
