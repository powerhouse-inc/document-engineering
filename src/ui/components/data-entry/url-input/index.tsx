export {
  UrlInput,
  type PlatformIcon,
  type UrlInputProps,
} from "./url-input.js";
