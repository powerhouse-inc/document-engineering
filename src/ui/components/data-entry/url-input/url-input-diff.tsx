import { FormGroup } from "../../../../scalars/components/fragments/form-group/form-group.js";
import { FormLabel } from "../../../../scalars/components/fragments/form-label/form-label.js";
import { SplittedInputDiff } from "../input/splitted-input-diff.js";
import type { UrlInputProps, UrlInputWithDifference } from "./url-input.js";

interface UrlInputDiffProps extends UrlInputWithDifference {
  value: string;
  label: UrlInputProps["label"];
  required: UrlInputProps["required"];
  platformIcons: UrlInputProps["platformIcons"];
}

const UrlInputDiff = ({
  value = "",
  label,
  required,
  viewMode,
  diffMode = "sentences",
  baseValue = "",
  platformIcons,
}: UrlInputDiffProps) => {
  return (
    <FormGroup>
      {label && (
        <FormLabel disabled={true} required={required}>
          {label}
        </FormLabel>
      )}
      <SplittedInputDiff
        baseValue={baseValue}
        value={value}
        viewMode={viewMode}
        diffMode={diffMode}
        asLink={true}
        platformIcons={platformIcons}
      />
    </FormGroup>
  );
};

export { UrlInputDiff };
