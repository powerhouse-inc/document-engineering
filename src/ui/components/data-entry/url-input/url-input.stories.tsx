import type { Meta, StoryObj } from '@storybook/react'
import {
  getDefaultArgTypes,
  getValidationArgTypes,
  PrebuiltArgTypes,
  StorybookControlCategory,
} from '../../../../scalars/lib/storybook-arg-types.js'
import { UrlInput } from './url-input.js'

/**
 * The `UrlInput` component provides an input field for web URLs.
 * It supports multiple configuration properties like:
 * - label
 * - description
 * - platformIcons
 * - showWarnings
 *
 * Features include:
 * - Platform-specific icons for known websites
 * - Automatic URL warnings
 * - Trailing space trimming
 *
 * > **Note:** This component does not have built-in validation. If you need built-in validation
 * > you can use the [UrlField](?path=/docs/scalars-url-field--readme)
 * > component.
 */

const meta: Meta<typeof UrlInput> = {
  title: 'Data Entry/Url Input',
  component: UrlInput,
  decorators: [
    (Story) => (
      <div style={{ width: '280px', margin: '1rem auto 0' }}>
        <Story />
      </div>
    ),
  ],
  parameters: {
    layout: 'padded',
    chromatic: {
      disableSnapshot: true,
    },
  },
  tags: ['autodocs'],
  argTypes: {
    ...getDefaultArgTypes(),
    ...PrebuiltArgTypes.placeholder,

    showWarnings: {
      control: 'boolean',
      description: 'Controls whether warnings are displayed to the user',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: 'true' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    platformIcons: {
      control: 'object',
      description: 'An object where the key is the hostname and the value is the icon name or a React element',
      table: {
        type: { summary: 'Record<string, IconName | ReactElement>' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    ...getValidationArgTypes({
      enabledArgTypes: {
        validators: false,
        showErrorOnBlur: false,
        showErrorOnChange: false,
      },
    }),

    ...PrebuiltArgTypes.viewMode,
    ...PrebuiltArgTypes.baseValue,
  },
  args: {
    name: 'url-input',
  },
}

export default meta
type Story = StoryObj<typeof UrlInput>

export const Default: Story = {
  args: {
    label: 'Website URL',
    placeholder: 'https://example.com',
  },
}

export const WithDescription: Story = {
  args: {
    label: 'Forum URL',
    description: 'Enter the URL to your online forum',
    placeholder: 'https://myforum.com',
  },
}

export const Required: Story = {
  args: {
    label: 'LinkedIn',
    required: true,
    placeholder: 'https://linkedin.com/in/username',
  },
}

export const WithDefaultValue: Story = {
  args: {
    label: 'With Default Value',
    defaultValue: 'https://example.com',
  },
}

export const Disabled: Story = {
  args: {
    label: 'Read Only URL',
    defaultValue: 'https://example.com',
    disabled: true,
  },
}

export const WithError: Story = {
  args: {
    label: 'Website URL',
    defaultValue: 'not-a-valid-url',
    errors: ['Please enter a valid URL'],
  },
}

export const WithWarning: Story = {
  args: {
    label: 'Website URL',
    defaultValue: 'https://example.com',
    warnings: ['URL may be unreachable'],
  },
}

export const WithPlatformIcon: Story = {
  args: {
    label: 'Website URL',
    defaultValue: 'https://github.com/test',
    platformIcons: {
      'github.com': 'Github',
      'linkedin.com': 'Linkedin',
      'twitter.com': 'XTwitter',
      'x.com': 'XTwitter',
      'youtube.com': 'Youtube',
      'forum.sky.money': 'Forum',
      'discord.com': 'Discord',
      'discord.gg': 'Discord',
    },
  },
}
