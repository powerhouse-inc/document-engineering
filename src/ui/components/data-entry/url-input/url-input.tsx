import React, { useCallback, useEffect, useId, useMemo, useState } from 'react'
import UrlFavicon from './url-favicon.js'
import { useURLWarnings } from './useURLWarnings.js'
import { FormDescription } from '../../../../scalars/components/fragments/form-description/index.js'
import { FormGroup } from '../../../../scalars/components/fragments/form-group/index.js'
import { FormLabel } from '../../../../scalars/components/fragments/form-label/index.js'
import { FormMessageList } from '../../../../scalars/components/fragments/form-message/index.js'
import { sharedValueTransformers } from '../../../../scalars/lib/shared-value-transformers.js'
import ValueTransformer from '../../../../scalars/components/fragments/value-transformer/index.js'
import { Input } from '../input/index.js'
import { UrlInputDiff } from './url-input-diff.js'
import { cn } from '../../../../scalars/lib/utils.js'
import type { UrlInputProps } from './types.js'

const UrlInput = React.forwardRef<HTMLInputElement, UrlInputProps>(
  (
    {
      label,
      description,
      required,
      disabled,
      showWarnings = true,
      warnings: warningsProp,
      errors,
      platformIcons,
      value,
      defaultValue,
      className,
      onBlur,
      onChange,
      onKeyDown,
      // diff props
      viewMode = 'edition',
      baseValue,
      ...props
    },
    ref
  ) => {
    const idGenerated = useId()
    const id = props.id ?? idGenerated
    const hasError = !!errors?.length

    const [internalValue, setInternalValue] = useState(value ?? defaultValue ?? '')
    const { warnings, checkForWarnings } = useURLWarnings(internalValue)

    const showIcon = Object.keys(platformIcons ?? {}).length > 0

    const combinedWarnings = useMemo(() => {
      return [...(warningsProp ?? []), ...warnings]
    }, [warningsProp, warnings])

    const handleBlur = useCallback(
      (event: React.FocusEvent<HTMLInputElement>) => {
        checkForWarnings()
        onBlur?.(event)
      },
      [checkForWarnings, onBlur]
    )

    const handleChange = useCallback(
      (event: React.ChangeEvent<HTMLInputElement>) => {
        setInternalValue(event.target.value)
        onChange?.(event)
      },
      [onChange]
    )

    const handleWarningsOnEnter = useCallback(
      (event: React.KeyboardEvent<HTMLInputElement>) => {
        if (event.key === 'Enter') {
          checkForWarnings()
        }
        onKeyDown?.(event)
      },
      [checkForWarnings, onKeyDown]
    )

    // prevent url from having trailing spaces
    const transformers = useMemo(() => [sharedValueTransformers.trimOnBlur()], [])

    useEffect(() => {
      if (value !== undefined) {
        setInternalValue(value)
      }
    }, [value])

    if (viewMode === 'edition') {
      return (
        <FormGroup>
          <FormLabel htmlFor={id} required={required} disabled={disabled} hasError={!!errors?.length}>
            {label}
          </FormLabel>
          <div className="relative">
            <ValueTransformer transformers={transformers}>
              <Input
                disabled={disabled}
                id={id}
                ref={ref}
                type="url"
                required={required}
                value={internalValue}
                onChange={handleChange}
                aria-invalid={hasError}
                className={cn(showIcon && 'pl-8', className)}
                data-cast="URLTrim"
                onBlur={handleBlur}
                onKeyDown={handleWarningsOnEnter}
                {...props}
              />
            </ValueTransformer>
            <UrlFavicon url={internalValue} platformIcons={platformIcons} className="absolute left-2.5 top-0" />
          </div>
          {description && <FormDescription>{description}</FormDescription>}
          {showWarnings && combinedWarnings.length > 0 && (
            <FormMessageList messages={combinedWarnings} type="warning" />
          )}
          {errors && <FormMessageList messages={errors} type="error" />}
        </FormGroup>
      )
    }

    return (
      <UrlInputDiff
        value={internalValue}
        label={label}
        required={required}
        viewMode={viewMode}
        baseValue={baseValue}
        platformIcons={platformIcons}
      />
    )
  }
)

UrlInput.displayName = 'UrlInput'

export { UrlInput }
