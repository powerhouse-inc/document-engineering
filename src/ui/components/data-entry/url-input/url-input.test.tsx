import { render, screen, waitFor } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { describe, expect, it } from 'vitest'
import { UrlInput } from './url-input.js'

describe('UrlInput', () => {
  it('should match snapshot', () => {
    const { container } = render(
      <UrlInput
        name="test-url"
        label="Website URL"
        description="Enter your website URL"
        placeholder="https://example.com"
        required
        warnings={['URL may be unreachable']}
        errors={['Invalid URL format']}
      />
    )

    expect(container).toMatchSnapshot()
  })

  it('should render with basic props', () => {
    render(<UrlInput data-testid="url-field" name="test-url" label="Website URL" placeholder="https://example.com" />)

    expect(screen.getByTestId('url-field')).toBeInTheDocument()
    expect(screen.getByPlaceholderText('https://example.com')).toBeInTheDocument()
  })

  it('should show required indicator when required', () => {
    render(<UrlInput data-testid="url-field" name="test-url" label="Website URL" required />)

    expect(screen.getByText('*')).toBeInTheDocument()
    expect(screen.getByTestId('url-field')).toHaveAttribute('required')
  })

  it('should show description when provided', () => {
    render(<UrlInput name="test-url" label="Website URL" description="Enter your website URL" />)

    expect(screen.getByText('Enter your website URL')).toBeInTheDocument()
  })

  it('should show error message when provided', async () => {
    render(<UrlInput name="test-url" label="Website URL" errors={['Invalid URL format']} />)

    await waitFor(() => {
      expect(screen.getByText('Invalid URL format')).toBeInTheDocument()
    })
  })

  it('should show warning message when provided', () => {
    render(<UrlInput name="test-url" label="Website URL" warnings={['URL may be unreachable']} />)

    expect(screen.getByText('URL may be unreachable')).toBeInTheDocument()
  })

  it('should be disabled when disabled prop is true', () => {
    render(<UrlInput data-testid="url-field" name="test-url" label="Website URL" disabled />)

    expect(screen.getByTestId('url-field')).toBeDisabled()
  })

  it('should show warning when the URL could be truncated', async () => {
    const user = userEvent.setup()
    render(<UrlInput data-testid="url-field" name="test-url" label="Website URL" />)

    const input = screen.getByTestId('url-field')
    await user.type(input, 'https://example.com/test...')
    await user.tab() // trigger blur

    expect(await screen.findByText('URL may be truncated')).toBeInTheDocument()
  })

  it('should not show warnings if showWarnings prop is false', async () => {
    const user = userEvent.setup()
    render(<UrlInput data-testid="url-field" name="test-url" label="Website URL" showWarnings={false} />)

    const input = screen.getByTestId('url-field')
    await user.type(input, 'https://example.com/test...')
    await user.tab() // trigger blur

    expect(screen.queryByText('URL may be truncated')).not.toBeInTheDocument()
  })

  it('should render platform built-in icon when URL matches configured hostname', () => {
    render(
      <UrlInput
        data-testid="url-field"
        name="test-url"
        label="Website URL"
        value="https://github.com/test"
        platformIcons={{
          'github.com': 'Github',
        }}
      />
    )

    expect(screen.getByTestId('icon-fallback')).toBeInTheDocument()
  })

  it('should render a custom icon when hostname has no configured icon', () => {
    render(
      <UrlInput
        data-testid="url-field"
        name="test-url"
        label="Website URL"
        value="https://github.com/test"
        platformIcons={{
          'github.com': <span data-testid="custom-icon">Custom Icon</span>,
        }}
      />
    )

    expect(screen.getByTestId('custom-icon')).toBeInTheDocument()
  })

  it('should not render any icon when platformIcons prop is not provided', () => {
    render(<UrlInput data-testid="url-field" name="test-url" label="Website URL" />)

    expect(screen.queryByTestId('icon-fallback')).not.toBeInTheDocument()
  })
})
