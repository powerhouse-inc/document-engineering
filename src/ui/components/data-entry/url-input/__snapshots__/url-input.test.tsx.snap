// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UrlInput > should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <label
      class="inline-flex items-center text-sm font-medium leading-4 group-hover:!text-red-900 dark:group-hover:!text-red-900 text-red-900 dark:text-red-900 mb-[3px]"
      for=":r0:"
    >
      Website URL
      <span
        class="ml-1 text-gray-800 group-hover:text-gray-900 dark:text-gray-400 dark:group-hover:text-slate-50 !text-red-900 group-hover:!text-red-900"
      >
        *
      </span>
    </label>
    <div
      class="relative"
    >
      <input
        aria-invalid="true"
        class="flex h-9 w-full rounded-md text-sm font-normal leading-5 text-gray-900 dark:text-gray-50 dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white px-3 py-2 font-sans placeholder:text-gray-500 dark:placeholder:text-gray-600 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-0 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 dark:focus:bg-charcoal-900 focus:bg-gray-50 disabled:cursor-not-allowed disabled:border-gray-300 disabled:bg-gray-50 disabled:text-gray-700 disabled:dark:border-charcoal-800 disabled:dark:bg-charcoal-900 disabled:dark:text-gray-300"
        data-cast="URLTrim"
        id=":r0:"
        name="test-url"
        placeholder="https://example.com"
        type="url"
        value=""
      />
    </div>
    <p
      class="font-sans text-sm font-normal leading-5 text-gray-600 dark:text-gray-500"
    >
      Enter your website URL
    </p>
    <p
      class="mb-0 inline-flex items-center text-xs font-medium text-yellow-900 dark:text-yellow-900"
      data-type="warning"
    >
      URL may be unreachable
    </p>
    <p
      class="mb-0 inline-flex items-center text-xs font-medium text-red-900 dark:text-red-900"
      data-type="error"
    >
      Invalid URL format
    </p>
  </div>
</div>
`;
